pdbtools/pdb_selaltloc.py

```diff
@@ -114,338 +120,206 @@ def check_input(args):
+        process the PDB file (or PDB lines), convert it to a list.
 
-    # uses the same function names in the loop below. However, depending
-    # on the input options, the functions used are different. One is
-    # specific for byocc=True, and other specific for occ char selection
-    flush_func_multi_residues = flush_resloc_occ if byocc else flush_resloc
+        >>> from pdbtools.pdb_selaltloc import run
+        >>> with('input.pdb', 'r') as fin:
+        >>>     processed_lines = list(run(fin))
 
-    flush_func_single_residues = \
-        flush_resloc_occ_same_residue if byocc else flush_resloc_id_same_residue
+        For more example see:
 
-    # defines records and terminators
+        >>> import pdbtools
+        >>> help(pdbtools)
+    """
     records = ('ATOM', 'HETATM', 'ANISOU')
-    terminators = ('TER', 'END', 'CONECT', 'END', 'ENDMDL')
+    terminators = ('TER', 'END', 'CONECT', 'END', 'ENDMDL', 'MODEL')
+
+    # register atom information
+    register = dict()
+
+    # register comment lines
+    others = []
+
+    # register current chain
+    chain = None
+    prev_chain = None
 
+    # keep record of the line number. This will be used to sort lines
+    # after selecting the desired alternative location
+    nline = 0
+
+    # the loop will collect information on the different atoms
+    # throughout the PDB file until a new chain or any terminal line is
+    # found. At that point, the collected information is flushed because
+    # all altlocs for that block have been defined.
     for line in fhandle:
+        nline += 1
 
         if line.startswith(records):
-            # captures the relevant parameters
-            altloc = line[16]
-            resname = line[17:20]
-            resnum = line[22:26].strip()
-
-            if is_another_altloc_group(
-                    altloc, prev_altloc, resnum, prev_resnum,
-                    resname, prev_resname, altloc_lines, res_per_loc):
-                # if we see the altloc group has changed, we should flush
-                # the lines observed for the previous altloc group
-
-                # uses "for loop" instead of "yield from" to maintain
-                # compatibility with older python version
-                if partial_altloc(altloc_lines):
-                    flush_func = flush_func_single_residues
-                else:
-                    flush_func = flush_func_multi_residues
 
-                for __line in flush_func(selloc=selloc, altloc_lines=altloc_lines):
-                    yield __line
+            # here resnum + insertion code are taken to identify
+            # different residues
+            resnum = line[22:27].strip()
+            atomname = line[12:16]
+            altloc = line[16]
+            chain = line[21:22]
 
-                altloc_lines = dict()
-                res_per_loc = dict()
+            # flush lines because we enter a new chain
+            if chain != prev_chain:
```

Comment:
I'd really avoid storing all residues in a chain, especially since this operation is residue based. You can trigger a flush when you see a new residue. That saves a lot of memory, not that it matters much because chains are tiny, but keeps with the idea of generators/iterators.

Reply:
Expanding on this a little, I wonder if you can simplify the overall logic by doing the following. All ATOM/HETATM/ANISOU records have the same identifiers (chain, resnum, inscode). You can record these are you iterate since they _must_ be in order. If they're not, it's really not our problem.

Once you hit a new residue, look back at the lines you accumulated for the residue. Track altlocs, if any in any of those lines. If there are any altlocs we want to keep/discard, do it then while flushing and ignoring those lines. This requires 3x iterations for a residue block if there are altlocs - once to store, once to process, once to flush - and you could cut this down to two by doing the processing as you store (basically keep a dict of {pdbname: {altloc: occupancy}}).

I think this _might_ be simpler than what you have and has the added benefit that you also filter ANISOU lines that belong to discarded altlocs. Let me know what you think.

Reply:
> I'd really avoid storing all residues in a chain, especially since this operation is residue based. You can trigger a flush when you see a new residue. That saves a lot of memory, not that it matters much because chains are tiny, but keeps with the idea of generators/iterators.

While this is formally true, it is also true that there are some cases where altlocs are organized by altloc instead of residues. [See test example](https://github.com/haddocking/pdb-tools/blob/f092906a985990661d1ec87b0efd0748e7f26289/tests/data/dummy_altloc2.pdb#L3-L34). I know this is not the *official* format but if we added it to tests it is because we have seen it at some point.

> Expanding on this a little (...)

You are right; we *may* optimize the code further, but I don't know if it can be simplified. In this PR, I aimed to reach the most robust algorithm, disregarding fine-tuned performance. I am positive there won't be memory issues because lines will be flushed on every `chain`, `model` or terminal line. Because `pdb-tools` is for PDBs and not for CIFs, there won't be more than 100k lines stored simultaneously in the dictionary; that's about 10MB of memory.

Considering the pressing nature of this issue, I suggest we merge this PR with `[FEATURE]` so all bugs are addressed, and then we work calmly on performance issues. I quickly measured the performance on a *large* PDB:

```bash
· time 'pdb_selaltloc 7bqx.pdb'
pdb_selaltloc 7bqx.pdb: command not found

real	0m0,110s
user	0m0,090s
sys	0m0,025s
```

What do you think? :wink: :tennis: